src/mjolnir/osmpbfparser.cc

```diff
@@ -285,42 +285,70 @@ Member::Member(Member&& other)
     : member_type(other.member_type), member_id(other.member_id), role(std::move(other.role)) {
 }
 
+// parser constructor
+Parser::Parser(int numThreads) : thread_pool(numThreads) {}
+
 void Parser::parse(std::ifstream& file, const Interest interest, Callback& callback) {
+  // Read file in the main thread and queue tasks for worker threads
+  read_file(file, interest, callback);
+
+  // Wait for all jobs to complete
+  thread_pool.WaitForAllJobs();
+}
+
+void Parser::read_file(std::ifstream& file, const Interest interest, Callback& callback) {
   char* buffer = new char[MAX_UNCOMPRESSED_BLOB_SIZE];
   char* unpack_buffer = new char[MAX_UNCOMPRESSED_BLOB_SIZE];
 
-  // start from the top
+  // Start from the top
   file.clear();
   file.seekg(0, std::ios::beg);
 
-  // while there is more to read
+  // While there is more to read
   while (!file.eof()) {
-    // grab the blob header
+    // Grab the blob header
     bool finished = false;
     BlobHeader header = read_header(buffer, file, finished);
-    // if we didnt hit the end
+    // If we didn't hit the end
     if (!finished) {
-      // grab the blob that goes with the blob header
+      // Grab the blob that goes with the blob header
       int32_t sz = read_blob(buffer, unpack_buffer, file, header);
-      // if its data parse it
-      if (header.type() == "OSMData") {
-        parse_primitive_block(unpack_buffer, sz, interest, callback);
-        // if its something other than a header
-      } else if (header.type() == "OSMHeader") {
-        parse_header_block(unpack_buffer, sz);
-      } else {
-        LOG_WARN("Unknown blob type: " + header.type());
-      }
+
+      // Create a task and queue it for processing
+      Task task = { buffer, unpack_buffer, header, sz };
+      thread_pool.QueueJob([this, task, interest, &callback]() {
+        this->process_task(task, interest, callback);
+      });
```

Comment:
Unfortunately the `callback` here get mutated in this function, creating a race condition - multiple threads will mutate the same `callback` simultaneously.